client/src/pages/ScanStore.tsx

```diff
@@ -47,39 +91,259 @@ function ScanStore() {
+      displayItems(items);
+    }
+  };
+
+  const displayItems = (extractedItems: Item[]) => {
+    for (let i = 0; i < extractedItems.length; ++i) {
+      if (idxMap[extractedItems[i].barcode] != undefined) {
+        cartItems[idxMap[extractedItems[i].barcode]].quantity += 1;
+      } else {
+        let newItem = emptyCartItem();
+        newItem.item = extractedItems[i];
+        newItem.quantity = 1;
+        cartItems.push(newItem);
+      }
+    }
+    setShoppingList(cartItems);
+  };
+
+  const updateShoppingListQuantity = (idx: number, quantity: number) => {
+    if (quantity <= 0) {
+      cartItems.splice(idx, 1);
+    } else {
+      cartItems[idx].quantity = quantity;
+    }
+    setShoppingList(cartItems);
+  };
+
+  const renderShoppingList = () => {
+    let UI_List = [];
+    for (let i = 0; i < shoppingList.length; ++i) {
+      UI_List.push(
+        <Grid key={"grid" + i} item xs={12}>
+          <ItemCard
+            cartItem={shoppingList[i]}
+            idx={i}
+            updateItemQuantity={updateShoppingListQuantity}
+          />
+        </Grid>
+      );
+    }
+    return UI_List;
+  };
+
+  const toggleCart = () => {
+    setShowCart(!showCart);
+  };
+
+  const updateIdxMapping = () => {
+    // Retrieve from State first
+    for (let i = 0; i < shoppingList.length; ++i) {
+      cartItems.push(shoppingList[i]);
+      const barcode: string = shoppingList[i].item.barcode;
+      idxMap[barcode] = i;
+    }
+  };
+
+  const makePayment = () => {
+    //Yiheng: This should be just an ID without exposing the contents of our Order
+    window.location.href =
+      "/receipt?id=TEST_ORDER&contents=" +
+      encodeURIComponent(JSON.stringify(shoppingList));
+  };
+
+  const processImageBarcode = async (img: HTMLImageElement) => {
+    const codeReader = new BrowserMultiFormatReader();
+    if (img != null) {
+      const result = await codeReader
+        .decodeFromImage(img)
+        .then((res: any) => res)
+        .catch((err: any) => {
+          console.log(err);
+          return "";
+        });
+      console.log("Processed image: " + result);
+      setDebugItem(result.text);
+    } else {
+      console.log("Cannot find element: uploadImgSrc");
+    }
+  };
+
+  const testBarcode = () => {
+    const img = document.getElementById("testImgSrc") as HTMLImageElement;
+    processImageBarcode(img);
+  };
+
+  //TODO When we update shoppingList, we should send update to server?
+
+  // When we change what image we uploaded, run extract barcode client-side
+  useEffect(() => {
+    if (uploadImg.mimeType) {
+      // Have something
+      const img = document.getElementById("uploadImgSrc") as HTMLImageElement;
+      processImageBarcode(img);
+    }
+  }, [uploadImg]);
+
+  // After rerendering of UI and React popping variables off our stack,
+  // reinitialize vars
+  // Also, race? Will this run before all other useEffects? What's the ordering of execution here?
```

Comment:
could you explain a bit more about your issue here?

Reply:
Unsure of order of execution between different useEffect hooks. We are using the no-dependency useEffect hook for initialization of variables upon re-rendering which leads to concerns of race if the hooks can be reordered.
#20 will resolve this issue.